src/AuthContext.tsx

```diff
@@ -6,7 +6,8 @@ import {
   AuthContextProps,
 } from './AuthContextInterface';
 
-export const AuthContext = React.createContext<AuthContextProps | null>(null);
+export const AuthContext = React.createContext<AuthContextProps>(undefined as unknown as AuthContextProps);
```

Comment:
Is there any other way than this type-casting?

Reply:
How about changing the `AuthContext` to this:

```suggestion
export const AuthContext = React.createContext<AuthContextProps | undefined>(undefined);
```

This is what it should've been all along, but instead of doing this:

```typescript
if (!context) {
```

We do this:

```typescript
if (!notEmpty(context)) {
```

and add a function like this?

```typescript
export function notEmpty<T>(value: T | undefined): value is T {
  return value !== undefined;
}
```

Would that work?

Reply:
Please lets keep this simple:
- The mentioned Formik library is doing it the same way. This library has lots of users
- The context can **not** be null when you are a child of the AuthProvider, if you call it outside, there if the `if + throw error`

-> No need to make it complicated!

Reply:
> Please lets keep this simple:
> 
> * The mentioned Formik library is doing it the same way. This library has lots of users
> * The context can **not** be null when you are a child of the AuthProvider, if you call it outside, there if the `if + throw error`
> 
> -> No need to make it complicated!

I agree, I prefer keeping it simple. However, this type-casting isn't the best practices of Typescript. Even though Formik is doing this; I feel we should try to find a way. 

If you don't there are any other way, I am fine with it. As a consequence of this, we could contribute back to Formik to improve their createContext :)

Reply:
it took some time, but now i understand your proposal. I like this:
```
export const AuthContext = React.createContext<AuthContextProps | undefined>(undefined);
```

What is the benefit of `if (!notEmpty(context))`, `if (!context)` or the more explicit `if (context === undefined)` would also work?

Reply:
btw we can diretly keep
`export const AuthContext = React.createContext<AuthContextProps | null>(null);`
and only doing adapting the function useAuth return type + if

Reply:
> it took some time, but now i understand your proposal. I like this:
> 
> ```
> export const AuthContext = React.createContext<AuthContextProps | undefined>(undefined);
> ```
> 
> What is the benefit of `if (!notEmpty(context))`, `if (!context)` or the more explicit `if (context === undefined)` would also work?

We need to make sure the type returned from `useAuth` is `AuthContextProps` and not `AuthContextProps | undefined`. We just need to make sure, and I didn't open the project to check. I wrote the `notEmpty` thing, as I'm sure the `: value T is` will ensure that.

> btw we can diretly keep
> `export const AuthContext = React.createContext<AuthContextProps | null>(null);`
> and only doing adapting the function useAuth return type + if

True that! I just prefer `undefined` instead of `null` 😅

Reply:
Ok, i adapted it!
No more typescript casting and using `null` instead of `undefined`

And as there was always the warning, that the function useAuth does not properly define a return type, i added one